spec/models/spree/product_spec.rb

```diff
@@ -0,0 +1,88 @@
+require 'spec_helper'
+
+module Spree
+  RSpec.describe Product do
```

Comment:
Yeah reading the decorator and this, my twitch would be to create a class/object to encapsulate this work that's not the product itself. This also helps with removing God objects which tend to rear their heads on User/Product/Order. Totally up to you though.